src/librbd/ImageCtx.h

```diff
@@ -61,6 +62,50 @@ namespace librbd {
   template <typename> class ResizeRequest;
   }
 
+  class ImageQosStatus {
+  public:
+
+    ImageQosStatus() : write_throttling_active(false), read_throttling_active(false) {}
+
+    ~ImageQosStatus() {}
+
+    utime_t get_prev_limit(bool is_write) const {
+      if (is_write) {
+        return prev_write_limit;
+      } else {
+        return prev_read_limit;
+      }
+    }
+
+    void set_prev_limit(bool is_write, utime_t limit) {
+      if (is_write) {
+        prev_write_limit = limit;
+      } else {
+        prev_read_limit = limit;
+      }
+    }
+
+    void set_throttling_flag(bool is_write, bool flag) {
+      if (is_write) {
+        write_throttling_active = flag;
+      } else {
+        read_throttling_active = flag;
+      }
+    }
+
+    bool get_throttling_flag(bool is_write) {
+      if (is_write) {
+        return write_throttling_active;
+      } else {
+        return read_throttling_active;
+      }
+    }
+
+  private:
+    utime_t prev_write_limit, prev_read_limit;
```

Comment:
I remove the ImageQosStatus class into the ImageRequestWQ.
In the ImageRequestWQ constructor, it can not determine whether read limit or write limit is enabled because the apply_metadata in ImageCtx is executed after it. 
So the simple way is just creating two instances of the ImageQosStatus in the constructor of ImageRequestWQ , one for read and one for write.